aptos-move/aptos-aggregator/src/aggregator_extension.rs

```diff
@@ -96,206 +127,442 @@ impl AggregatorID {
+
+                if cur_value < input {
+                    let underflow_value = expect_ok(ok_overflow(addition_deltavalue(
+                        input,
+                        delta.minus(),
+                        self.max_value,
+                    )))?;
+                    // if value overflowed, we don't need to record it
+                    if let Some(underflow_value) = underflow_value {
+                        self.get_mut_history()
+                            .unwrap()
+                            .record_underflow(underflow_value);
+                    }
+                    Ok(false)
+                } else {
+                    let new_delta =
+                        expect_ok(delta.add(&DeltaValue::Negative(input), self.max_value))?;
+                    self.state = AggregatorState::Delta {
+                        speculative_start_value,
+                        delta: new_delta,
+                        history,
+                    };
+                    self.get_mut_history().unwrap().record_success(new_delta);
+                    Ok(true)
+                }
             },
         }
-
-        // Record side-effects of addition in history.
-        self.record();
-        Ok(())
     }
 
-    /// Implements logic for reading the value of an aggregator. As a
-    /// result, the aggregator knows it value (i.e. its state changes to
-    /// `Data`).
-    pub fn read_and_materialize(
+    /// Implements logic for doing a "cheap read" of an aggregator.
+    /// This means that we query the MVHashmap for the last committed value
+    /// of the aggregator, and store it in the `speculative_start_value` with
+    /// `LastCommittedValue` variant.
+    pub fn read_last_committed_aggregator_value(
         &mut self,
         resolver: &dyn AggregatorResolver,
-        id: &AggregatorID,
     ) -> PartialVMResult<u128> {
-        // If aggregator has already been read, return immediately.
-        if self.state == AggregatorState::Data {
-            return Ok(self.value);
+        match self.state {
+            AggregatorState::Data { value } => {
+                // If aggregator knows the value, return it.
+                Ok(value)
+            },
+            AggregatorState::Delta {
+                speculative_start_value,
+                delta,
+                history,
+            } => {
+                // If we performed a "cheap read" or "expensive read" operation before, use it.
+                if speculative_start_value != SpeculativeStartValue::Unset {
+                    return Ok(addition_deltavalue(
+                        speculative_start_value.get_value(),
+                        delta,
+                        self.max_value,
+                    )?);
+                }
+                // Otherwise, we have to go to storage and read the value.
+                let value_from_storage = resolver
+                    .resolve_aggregator_value(&self.id, AggregatorReadMode::Speculative)
```

Comment:
reading the last committed is not speculative? Should it be Mode::Precise here?

Reply:
Nvm, I see the second function. To me committed sounds pretty non-speculative 😂 and most recent sounds speculative instead